### DISCUSSION TITLE
Is it possible to "interleave" patterns  in Strudel?

### DISCUSSION THREAD

**Discussion by mystery-house** [tidalcycles/strudel#1201] [Q&A] [answered]
My first live coding experience was with [FoxDot](https://github.com/Qirky/FoxDot), which approaches things a little differently. Strudel is absolutely fantastically amazing, but one of the things I've been missing and trying to replicate from FoxDot is the way that you can play sounds with arrays for pitch, duration, (and other modifiers), for example:

`p1 >> pluck([0,2, [3, 4]], dur=[1,.5])`
 
When executed, FoxDot recursively interleaves the inputs into a longer pattern with every possible combination of pitch and duration; so with the above structure you'd wind up with a sequence like:

pitch | duration
--- | ---
0 | 1
2 | .5
3 | 1
0 | .5
2 | 1
4 | .5

The arrays can be different lengths and each array can nest other arrays, which makes it easy to quickly create and tweak fantastically complex melodies rhythms.

I know you can pass multiple values to `slow()` and `fast()` in Strudel, but that seems to apply the values to the entire cycle rather than combining individual notes and durations.

I'm still in the "don't know what I don't know" learning phase with Strudel, so I'm not sure if this is conceptual roadblock on my part (maybe not possible with mini notation, but doable with api functions?), or if Strudel just doesn't work this way.

**Comment by mystery-house** [comment]
...Still don't know if there's an existing and/or more idiomatic way to do this, but yes, it is possible! 

I haven't published an official package yet but I wrote some JS functions to deal with the recursive permutation part, and I was able to import that into the REPL and map the results through `m()`.   

```
let pitches = [0,  2, [3,  4, ['~', 5]]];
let speeds = [1, 1, [.5, .25]];
let p = interleaveRecursive.interleaveConcat(2, '!', pitches, speeds )
let mini = p.map(item=>m(item))
```

https://strudel.cc/?q8xwXV6dWVe3

**Reply by felixroos** [reply]
another tip: when you prepend your fetch with await, you won't get an error on the first evaluation

**Reply by mystery-house** [reply]
(Actually it turned out the issue was trying to reference the imported var before the newly injected `script` element had also finished loading 😅)

**Comment by felixroos** [comment, marked as answer]
hey @mystery-house ! doing stuff like this is really a bit difficult with the way patterns work.. these list based sequencing techniques are very easy to implement when you have state (e.g. a pointer to the currently active element). Patterns don't have any state, which is why it's harder to represent something like this. It might still be possible, if you're able to find a pure/stateless function that returns the events for a certain time frame. Of course, precalculating is also an option, and probably sufficient in most cases.

On that note, it might be better if you'd not generate a mini notation string, but instead throw your events into [timeCat](https://strudel.cc/learn/factories/#s_cat), which allows you to specify events with durations. This way, it might also be possible to get proper highlighting

**Reply by mystery-house** [reply]
Thanks for confirming that it's a "just not really how cycles work" issue, and the pointer to timeCat. Lots to experiment with.

Marking answered with a link to your #646, which touches on the matter of state (and a good example with highlighting!)
